### PR TITLE
gh-102192: Replace PyErr_Fetch/Restore etc by more efficient alternatives (in Python/)

### PR DESCRIPTION
**Pull request by iritkatriel**
This PR does most of Python/.

* Issue: gh-102192

### CODE REVIEW

**Comment by iritkatriel** [comment]
Skipping news because these are internal-only changes, and the fact of PyErr_Fetch/Restore being obsolete is already in the news.

**Comment by iritkatriel** [comment]
There's a refcount issue somewhere. Checking.

File: Objects/exceptions.c [hunk 1]
```diff
@@ -3852,28 +3848,18 @@ _PyErr_TrySetFromCause(const char *format, ...)
...
-    if (tb != NULL) {
-        PyException_SetTraceback(val, tb);
-        Py_DECREF(tb);
-    }
-
     va_start(vargs, format);
     msg_prefix = PyUnicode_FromFormatV(format, vargs);
     va_end(vargs);
     if (msg_prefix == NULL) {
-        Py_DECREF(exc);
```
Comment (erlend-aasland) [inline Objects/exceptions.c]: We should still decref `exc` here, right? AFAICS, we still hold a strong reference to it (`PyException_SetCause` swallows it just below).

**Comment by erlend-aasland** [comment]
I must say this is a great improvement in readability. The new API is indeed a better API. Thanks!

**Comment by iritkatriel** [comment]
There's only one failure now, in an AMD64 RHEL7 builedbot which has been flaky for a couple of months. I'll call it a pass.

**Comment by iritkatriel** [comment]
> There's only one failure now, in an AMD64 RHEL7 builedbot 

It passes now, so this PR should be fine.

File: Python/ceval.c [hunk 2]
```diff
@@ -13,7 +13,7 @@
 #include "pycore_object.h"        // _PyObject_GC_TRACK()
 #include "pycore_moduleobject.h"  // PyModuleObject
 #include "pycore_opcode.h"        // EXTRA_CASES
-#include "pycore_pyerrors.h"      // _PyErr_Fetch()
+#include "pycore_pyerrors.h"      // _PyErr_Fetch(), _PyErr_GetRaisedException()
```
Comment (markshannon) [inline Python/ceval.c:16]: Remove the `_PyErr_Fetch` from the comment?

**Review by markshannon — CHANGES_REQUESTED** [review]
One minor edit needed, otherwise LGTM.

File: Python/ceval.c (hunk 2 above)
Comment (iritkatriel) [reply Python/ceval.c:16]: There still are a couple left in the tracing stuff (the trace function wants the triplet).

**Review by markshannon — APPROVED** [review]

**Pull request merged**